### ISSUE TITLE
BN should know about most used basic data types for each platform

### ISSUE THREAD

**Issue by op2786** [Vector35/binaryninja-api#3836]
When I reverse windows binaries, I expect binary ninja to know about what DWORD, ULONG, PVOID, PWCHAR etc is. Currently it does not know about them automatically. Most of the time api functions uses that basic types, and I want to see them in my decompiler output. I think this is critical to speed up reverse engineering because working with the most used basic data types on the aforementioned platform can save our brain from wasting a few extra ATPs.

**Comment by FuzzySecurity** [comment]
^ this is a big one for binja on Windows targets at least

**Comment by fuzyll** [comment]
We have types like `DWORD` defined in our type libraries. If your binary loads `msvcrt.dll` as part of its normal load process, this should cause a type like `DWORD` be loaded automatically into your `.bndb`.

In some cases, this doesn't happen. To load the type manually, you can do something like:

```
tl = bv.platform.get_type_libraries_by_name("msvcrt.dll")
dword = bv.import_library_type("DWORD", tl[0])
bv.define_user_type("DWORD", dword)
```

Note that there are a bunch of limitations here, though. For one, you have to be able to actually get a handle to the `msvcrt.dll` type library, and I'm not actually sure how to do that if it doesn't automatically get loaded into `bv.type_libraries` or `bv.platform.type_libraries`. So, if you want to use `DWORD` in an ARM64 binary, I think you're still out of luck. The same goes for if you want to use `DWORD` in a macOS or Linux binary.

**Comment by FuzzySecurity** [comment]
Some kind of lookup mechanism would be ideal where it can poll a symbols from all data sources. The type support and lookup works really well in Ghidra btw. In the list you could display duplicate symbol defs from different data sources also.

![image](https://user-images.githubusercontent.com/10997340/222000430-819b0af0-8e21-45a7-9391-f743ca77cca4.png)

**Comment by fuzyll** [comment]
Agreed! But, _that_ issue is #3829. This issue tracks the fact that we are missing some very common data types for certain platforms. _That_ issue is for making them have a vastly improved UX. 🙂

**Comment by FuzzySecurity** [comment]
I think they are sort of the same issue. I agree that the way this is displayed to a user is a different matter (for the UI design/usability team). My comment was about the proposed solution of using some code to extract a type out of a library manually if it is not found. I think that is something binja should take care of under-the-hood when the user tries to re-type something. In the case that a symbol exists in multiple libraries then the user can be prompted which symbol to apply, the issue here is that this lookup doesn't happen and so the symbol is not found. Maybe that requires on-demand loading, I don't know about the internals.

**Comment by plafosse** [comment]
I've got a branch I'm working on that will have all the basic windows types as well as updated type libraries.

**Comment by plafosse** [comment]
This has been resolved in 3.6.4684
